domain/info.go

```diff
@@ -0,0 +1,174 @@
+	return &InfoSyncer{
+		etcdCli:        etcdCli,
+		info:           getServerInfo(id),
+		serverInfoPath: fmt.Sprintf("%s/%s", ServerInformationPath, id),
+	}
+}
+
+func getServerInfo(id string) *ServerInfo {
+	cfg := config.GetGlobalConfig()
+	info := &ServerInfo{
+		ID:         id,
+		IP:         cfg.AdvertiseAddress,
+		Port:       cfg.Port,
+		StatusPort: cfg.Status.StatusPort,
+		Lease:      cfg.Lease,
+	}
+	info.Version = mysql.ServerVersion
+	info.GitHash = printer.TiDBGitHash
+	return info
+}
+
+//GetServerInfo gets self server static information.
+func (is *InfoSyncer) GetServerInfo() *ServerInfo {
+	return is.info
+}
+
+// GetServerInfoByID gets owner server static information from Etcd.
+func (is *InfoSyncer) GetServerInfoByID(ctx context.Context, id string) (*ServerInfo, error) {
+	if is.etcdCli == nil || id == is.info.ID {
+		return is.info, nil
+	}
+	key := fmt.Sprintf("%s/%s", ServerInformationPath, id)
+	infoMap, err := getInfo(ctx, is.etcdCli, key)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	info, ok := infoMap[id]
+	if !ok {
+		return nil, errors.Errorf("[infoSyncer] get %s failed", key)
+	}
+	return info, nil
+}
+
+// GetAllServerInfo gets all servers static information from Etcd.
+func (is *InfoSyncer) GetAllServerInfo(ctx context.Context) (map[string]*ServerInfo, error) {
+	allInfo := make(map[string]*ServerInfo)
+	if is.etcdCli == nil {
+		allInfo[is.info.ID] = is.info
+		return allInfo, nil
+	}
+	allInfo, err := getInfo(ctx, is.etcdCli, ServerInformationPath, clientv3.WithPrefix())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return allInfo, nil
+}
+
+// StoreServerInfo stores self server static information to Etcd when domain Init.
+func (is *InfoSyncer) StoreServerInfo(ctx context.Context) error {
+	if is.etcdCli == nil {
+		return nil
+	}
+	infoBuf, err := json.Marshal(is.info)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = ddl.PutKVToEtcd(ctx, is.etcdCli, ddl.KeyOpDefaultRetryCnt, is.serverInfoPath, hack.String(infoBuf))
+	return errors.Trace(err)
+}
+
+// RemoveServerInfo remove self server static information from Etcd.
+func (is *InfoSyncer) RemoveServerInfo() {
+	if is.etcdCli == nil {
+		return
+	}
+	err := ddl.DeleteKeyFromEtcd(is.serverInfoPath, is.etcdCli)
+	if err != nil {
+		log.Errorf("[infoSyncer] remove self server info failed %v", err)
+	}
+}
+
+// getInfo gets server information from Etcd according to the key and opts.
+func getInfo(ctx context.Context, etcdCli *clientv3.Client, key string, opts ...clientv3.OpOption) (map[string]*ServerInfo, error) {
+	var err error
+	allInfo := make(map[string]*ServerInfo)
+	for {
+		select {
+		case <-ctx.Done():
+			err = errors.Trace(ctx.Err())
+			return nil, err
+		default:
+		}
+		childCtx, cancel := context.WithTimeout(ctx, ddl.KeyOpDefaultTimeout)
+		resp, err := etcdCli.Get(childCtx, key, opts...)
+		cancel()
+		if err != nil {
+			log.Infof("[infoSyncer] get %s failed %v, continue checking.", key, err)
```

Comment:
How many times it will try?
Will there be too many logs?

Reply:
I will add retry count and timeout.